File: camunda-carbon-reductor-c7/src/test/java/de/envite/greenbpm/carbonreductorconnector/ElementTemplateTest.java
package de.envite.greenbpm.carbonreductorconnector;

import org.camunda.bpm.engine.runtime.ProcessInstance;
import org.camunda.bpm.engine.test.Deployment;
import org.camunda.bpm.engine.test.junit5.ProcessEngineExtension;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import java.util.Map;

import static org.assertj.core.api.Assertions.*;
import static org.camunda.bpm.engine.test.assertions.ProcessEngineTests.*;

@ExtendWith(ProcessEngineExtension.class)
class ElementTemplateTest {
    private static final String PROCESS_KEY = "CarbonReductorTestProcess";
    private static final String CARBON_REDUCTOR_ID = "ST_Carbon_Reductor";

    @Test
    @Deployment(resources = "test.bpmn")
    void test() {
        ProcessInstance processInstance = runtimeService().startProcessInstanceByKey(PROCESS_KEY);
        assertThat(processInstance).isStarted();
        assertThat(processInstance).isWaitingAt(CARBON_REDUCTOR_ID);
        // input variables
        Map<String, Object> inputVariables = runtimeService().getVariables(externalTask().getExecutionId());
        assertThat(inputVariables).containsOnlyKeys(
                "remainingProcessDuration", "maximumProcessDuration", "location", "milestone", "errorHandling", "measurementOnly", "thresholdValue", "thresholdEnabled");
        complete(externalTask(CARBON_REDUCTOR_ID), Map.of(
                "originalCarbon", 100,
                "actualCarbon", 50,
                "savedCarbon", 50,
                "reducedCarbon", 50,
                "executionDelayed", true,
                "delayedBy", 100000));
        // output variables
        assertThat(processInstance).hasVariables(
                "originalCarbonMapped",
                "actualCarbonMapped",
                "savedCarbonMapped",
                "reducedCarbonMapped",
                "executionDelayedMapped",
                "delayedByMapped");
        assertThat(processInstance).hasVariables(
                "originalCarbon",
                "actualCarbon",
                "savedCarbon",
                "reducedCarbon",
                "executionDelayed",
                "delayedBy");
        assertThat(processInstance).isEnded();
    }
}

File: camunda-carbon-reductor-c7/src/test/kotlin/de/envite/greenbpm/carbonreductorconnector/ElementTemplateTest.kt
package de.envite.greenbpm.carbonreductorconnector

import org.camunda.bpm.engine.test.Deployment
import org.camunda.bpm.engine.test.assertions.bpmn.BpmnAwareTests.runtimeService
import org.camunda.bpm.engine.test.assertions.bpmn.BpmnAwareTests.assertThat
import org.camunda.bpm.engine.test.assertions.bpmn.BpmnAwareTests.complete
import org.camunda.bpm.engine.test.assertions.bpmn.BpmnAwareTests.externalTask
import org.camunda.bpm.engine.test.junit5.ProcessEngineExtension
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith

@ExtendWith(ProcessEngineExtension::class)
class ElementTemplateTest {
    private val processKey = "CarbonReductorTestProcess"
    private val carbonReductorId = "ST_Carbon_Reductor"

    @Test
    @Deployment(resources = ["test.bpmn"])
    fun test() {
        val processInstance = runtimeService().startProcessInstanceByKey(processKey)
        assertThat(processInstance).isStarted
        assertThat(processInstance).isWaitingAt(carbonReductorId)
        complete(
            externalTask(carbonReductorId), mapOf<String, Any>(
                "originalCarbon" to 100,
                "actualCarbon" to 50,
                "savedCarbon" to 50,
                "reducedCarbon" to 50,
                "executionDelayed" to true,
                "delayedBy" to 100000
            )
        )
        // input variables
        assertThat(processInstance).hasVariables(
            "remainingProcessDuration",
            "maximumProcessDuration",
            "carbonReductorMode",
            "location",
            "milestone"
        )
        // output variables
        assertThat(processInstance).hasVariables(
            "originalCarbonMapped",
            "actualCarbonMapped",
            "savedCarbonMapped",
            "reducedCarbonMapped",
            "executionDelayedMapped",
            "delayedByMapped"
        )
        assertThat(processInstance).hasVariables(
            "originalCarbon",
            "actualCarbon",
            "savedCarbon",
            "reducedCarbon",
            "executionDelayed",
            "delayedBy"
        )
        assertThat(processInstance).isEnded
    }
}